### PR TITLE
Fix #29: preserve comments and regions (trivia) during fixes

### LINKED ISSUE #29: Comments and Regions Are Removed If the Order of ExpectedException Comes Before Test
It appears that the Analzyer Fixer will remove Comments and Regions if the `ExpectedException` attribute is decorated ABOVE the `Test` Attribute

Consider the Following Test:

```csharp
using NUnit.Framework;
using System;
[TestFixture]
public class TestClass
{
    /// <summary>
    /// There was a comment here
    /// </summary>
    [ExpectedException(typeof(Exception))]
    [Test]
    public void TestMethod()
    {
        throw new Exception();
    }
}
```

After running the after running the Analzyer Fix the Comment has been destroyed:
```csharp
using NUnit.Framework;
using System;
[TestFixture]
public class TestClass
{
    [Test]
    public void TestMethod()
    {
        Assert.Throws<Exception>(() =>
        {
            throw new Exception();
        });
    }
}
```

This will also affect if you had regions:
```csharp
using NUnit.Framework;
using System;
[TestFixture]
public class TestClass
{
    #region Test
    [ExpectedException(typeof(Exception))]
    [Test]
    public void TestMethod()
    {
        throw new Exception();
    }
    #endregion
}
```

After running the Analyzer Fix:
```csharp

using NUnit.Framework;
using System;
[TestFixture]
public class TestClass
{
    [Test]
    public void TestMethod()
    {
        Assert.Throws<Exception>(() =>
        {
            throw new Exception();
        });
    }
    #endregion
}
```

This last one will at least throw a compiler error because the `#endregion` has no `#region` starting directive.

### PR DESCRIPTION
**Pull request by aolszowka**
I tried your suggestion but it causes the other tests to fail (looks like its no longer adding a whitespace?)

### CODE REVIEW

**Comment by wachulski** [comment]
@aolszowka do you need any help/guidance with the failing tests?

**Comment by wachulski** [comment]
Ah, I see, if there is a method split (into 2 or more exception specific ones) then the second and all the next are missing a single leading empty line (as they borrow leading/trailing whitespace from the original method).

Let me see what I can do for it.

**Comment by wachulski** [comment]
Ok @aolszowka. I did a few things. It just needs final polishing touches (which I leave to you).

1. I rebased on the master in a GIT proper way. I see you managed to merge the master into your `aolszowka:comment-and-region-truncation` in some mysterious way that it wasn't marked as a merge commit - and this we should avoid. I kindly encourage you to read about `git merge` and `git rebase` utilities.
2. You can just reset your branch to mine (`wachulski/comment-and-region-truncation`). You can delete and checkout again if you don't know how to do it.
3. As you can see I left your unit tests and provided a fix for the issue. All unit tests should go green now.
4. What's left are tests. This is that final touch:

- please extract your tests into a fixture like: `ExceptionExpectancy/TriviaFixture.cs`
- rename according to the convention: `WhenSthIsSetup_PreservesThisAndThat` e.g. `CommentsArePreserved_ExpectedExceptionAttributeBelowTest` turns into `WhenExpectedExceptionAttributeBelowTest_FixesAndCommentsArePreserved` (and you see it lacks information about the comments arranged for the test, just some info about attribute placement, but should be telling about comments instead/too)
- please also add some tests for TestCases with various exceptions (this will produce 1 method split into several ones) [I'm particularly interested if such splits preserve comments/regions correctly]

@aolszowka Are you fancy continuing? In short, we're very close to fix it, merge and release 1.1.1.

**Comment by aolszowka** [comment]
Hey @wachulski sorry the holiday rush has started here before we have time off; I'll see if I can address this today.

> 1.  rebased on the master in a GIT proper way.

Yah I was struggling to understand how to correctly resync to the master branch; I'll take a look at the docs again; do you have any videos you recommend to learn how to do this properly? It seems like I was unable to do it from the Web Interface, is the CLI the only way?

> 2. you can delete and checkout again if you don't know how to do it.

I'll probably do that this time around.

Again let me see if I can knock this out today; tomorrow at the latest.

**Comment by wachulski** [comment]
No rush 😃 thx for your contribution and patience so far. If my tone was a bit harsh and hurrying up - excuse me, I didn't mean that.

You contributed quite a lot actually and I didn't want to finalize everything myself. So, final touches left for you 😃 

When it comes to GIT, the best resource I learnt from is this book (available online, free): https://git-scm.com/book/en/v2 I also see that there are plenty of other tutorials atm, e.g. https://try.github.io/ by GH itself.

If you're Windows-based then I also strongly recommend some visual tool (e.g. https://github.com/gitextensions/gitextensions which I use heavily).

**Comment by aolszowka** [comment]
Hey I am on it right now; I am a little confused by this statement:

>please also add some tests for TestCases with various exceptions (this will produce 1 method split into several ones) [I'm particularly interested if such splits preserve comments/regions correctly]

I am unfamiliar what what this means?

At first I thought you maybe meant having multiple tests in there; but upon reading it again I do not think that is what you intended?

```csharp
        [Test]
        public void WhenExpectedExceptionAttributeIsBelowMultipleTests_CommentsArePreserved()
        {
            var source = @"
using NUnit.Framework;
using System;

[TestFixture]
public class TestClass
{
    /// <summary>
    /// There was another comment here
    /// </summary>
    [Test]
    [ExpectedException(typeof(InvalidOperationException))]
    public void TestMethod2()
    {
        throw new InvalidOperationException();
    }

    /// <summary>
    /// There was a comment here
    /// </summary>
    [Test]
    [ExpectedException(typeof(Exception))]
    public void TestMethod()
    {
        throw new Exception();
    }
}";

            var expected = @"
using NUnit.Framework;
using System;

[TestFixture]
public class TestClass
{
    /// <summary>
    /// There was another comment here
    /// </summary>
    [Test]
    public void TestMethod2()
    {
        Assert.Throws<InvalidOperationException>(() =>
        {
            throw new InvalidOperationException();
        });
    }

    /// <summary>
    /// There was a comment here
    /// </summary>
    [Test]
    public void TestMethod()
    {
        Assert.Throws<Exception>(() =>
        {
            throw new Exception();
        });
    }
}";

            VerifyCSharpFix(source, expected);
        }
```

**Comment by wachulski** [comment]
Take a look at tests in `MixedExpectedExceptionAndTestCasesFixture` and `TestCaseOnlyFixture` - tests that result in 1 method split into several ones. Like: https://github.com/wachulski/nunit-migrator/blob/76c807337e8c1804181b0e245b0ca913f69d80b1/nunit.migrator.tests/ExceptionExpectancy/TestCaseOnlyFixture.cs#L82

**Comment by aolszowka** [comment]
Ok that makes sense; and bummer it does not work; would you like me to commit the test? I think I'll need to create a new pull request.

```csharp
        [Test]
        public void For2TCs_WhenExpectedExceptionAttribute_CommentsArePreserved()
        {
            var source = @"
using NUnit.Framework;
using System;

[TestFixture]
public class TestClass
{
    /// <summary>
    /// There was a comment here
    /// </summary>
    /// <param name=""input"">Comment for input</param>
    /// <remarks>This was a remark</remarks>
    [TestCase(1, ExpectedException = typeof(Exception))]
    [TestCase(2)]
    public void TestMethod( int input )
    {
        throw new Exception();
    }
}";

            var expected = @"
using NUnit.Framework;
using System;

[TestFixture]
public class TestClass
{
    /// <summary>
    /// There was a comment here
    /// </summary>
    /// <param name=""input"">Comment for input</param>
    /// <remarks>This was a remark</remarks>
    [TestCase(1)]
    public void TestMethod( int input )
    {
        Assert.Throws<Exception>(() =>
        {
            throw new Exception();
        });
    }

    /// <summary>
    /// There was a comment here
    /// </summary>
    /// <param name=""input"">Comment for input</param>
    /// <remarks>This was a remark</remarks>
    [TestCase(2)]
    public void TestMethod( int input )
    {
        throw new Exception();
    }
}";

            VerifyCSharpFix(source, expected);
        }
    }
```

The result ends up being that the comments are only pulled into the first test case. In addition I would have expected test case 1 to remain at the "top" but instead it was moved down, see the diff to see what I'm talking about:

image


This test also showed some interesting behavior; it ended up appending a _ShouldThrowException; which was not expected in my mind; but in this case the comments were preserved to both test cases (see the attached diff):

```csharp
        [Test]
        public void For2TCs_WhenExpectedExceptionIsSecondTestCase_CommentsArePreserved()
        {
            var source = @"
using NUnit.Framework;
using System;

[TestFixture]
public class TestClass
{
    /// <summary>
    /// There was a comment here
    /// </summary>
    /// <param name=""input"">Comment for input</param>
    /// <remarks>This was a remark</remarks>
    [TestCase(1)]
    [TestCase(2, ExpectedException = typeof(Exception))]
    public void TestMethod( int input )
    {
        throw new Exception();
    }
}";

            var expected = @"
using NUnit.Framework;
using System;

[TestFixture]
public class TestClass
{
    /// <summary>
    /// There was a comment here
    /// </summary>
    /// <param name=""input"">Comment for input</param>
    /// <remarks>This was a remark</remarks>
    [TestCase(1)]
    public void TestMethod( int input )
    {
        throw new Exception();
    }

    /// <summary>
    /// There was a comment here
    /// </summary>
    /// <param name=""input"">Comment for input</param>
    /// <remarks>This was a remark</remarks>
    [TestCase(2)]
    public void TestMethod( int input )
    {
        Assert.Throws<Exception>(() =>
        {
            throw new Exception();
        });
    }
}";

            VerifyCSharpFix(source, expected);
        }
```
image

**Comment by aolszowka** [comment]
Regions do not fair well either; in this case it will produce code that will not compile (it will throw CS1038 because there is no #endregion).

Test Case:

```csharp
        [Test]
        public void For2TCs_WhenExpectedException_RegionsArePreserved()
        {
            var source = @"
using NUnit.Framework;
using System;

[TestFixture]
public class TestClass
{
    #region TestRegion
    [TestCase(1)]
    [TestCase(2, ExpectedException = typeof(Exception))]
    public void TestMethod(int input)
    {
        throw new Exception();
    }
    #endregion
}";

            var expected = @"
using NUnit.Framework;
using System;

[TestFixture]
public class TestClass
{
    #region TestRegion
    [TestCase(1)]
    public void TestMethod(int input)
    {
        throw new Exception();
    }

    [TestCase(2)]
    public void TestMethod(int input)
    {
        Assert.Throws<Exception>(() =>
        {
            throw new Exception();
        });
    }
    #endregion
}";

            VerifyCSharpFix(source, expected);
        }
```

Result:
image

**Comment by wachulski** [comment]
I've just pushed another fix, this time the one accounting for mixed exception and non-exception test cases. https://github.com/wachulski/nunit-migrator/tree/comment-and-region-truncation should be ok.

Are you fine with the changes? Do you see any other potential problem here? When it comes to PR, we don't need a new one. Just reset your branch to mine and push with `--force`.

**Comment by aolszowka** [comment]
Do we expect the comments to be duplicated when the test case is extracted? I am OK with them not being duplicated (as it is with your most current fix) with that but maybe we should add a warning about it?

Just so we're clear I'm talking about `For2TCs_WhenExpectedExceptionIsFirstTestCase_CommentsArePreserved()`

Here is the source:

```csharp
using NUnit.Framework;
using System;

[TestFixture]
public class TestClass
{
    /// <summary>
    /// There was a comment here
    /// </summary>
    /// <param name="input">Comment for input</param>
    /// <remarks>This was a remark</remarks>
    [TestCase(1, ExpectedException = typeof(Exception))]
    [TestCase(2)]
    public void TestMethod( int input )
    {
        throw new Exception();
    }
}
```

I could see it being transformed two ways:

1. This is the way it is working after your fixes (note the comment is missing from the new test case):

```csharp
using NUnit.Framework;
using System;

[TestFixture]
public class TestClass
{
    /// <summary>
    /// There was a comment here
    /// </summary>
    /// <param name="input">Comment for input</param>
    /// <remarks>This was a remark</remarks>
    [TestCase(2)]
    public void TestMethod( int input )
    {
        throw new Exception();
    }

    [TestCase(1)]
    public void TestMethod_ShouldThrowException( int input )
    {
        Assert.Throws<Exception>(() =>
        {
            throw new Exception();
        });
    }
}
```

2. This is another potential way it could have transformed the source (note that the comment is duplicated):
```csharp
using NUnit.Framework;
using System;

[TestFixture]
public class TestClass
{
    /// <summary>
    /// There was a comment here
    /// </summary>
    /// <param name="input">Comment for input</param>
    /// <remarks>This was a remark</remarks>
    [TestCase(2)]
    public void TestMethod( int input )
    {
        throw new Exception();
    }

    /// <summary>
    /// There was a comment here
    /// </summary>
    /// <param name="input">Comment for input</param>
    /// <remarks>This was a remark</remarks>
    [TestCase(1)]
    public void TestMethod_ShouldThrowException( int input )
    {
        Assert.Throws<Exception>(() =>
        {
            throw new Exception();
        });
    }
}
```


Thank you so much for your project and helping out!

Also thank you for being patient with me as I learn Git (coming from a Subversion background so things are slightly different). I found this really helpful https://help.github.com/articles/merging-an-upstream-repository-into-your-fork/

**Comment by wachulski** [comment]
When it comes to copying summary docs, I'd like to avoid it for several reasons:

* it adds complexity to trivia processing, filtering (to get only summary docs and not regular comments), taking into account whitespacing etc. (this risks breaking builds if done wrong)
* having two unit tests with same summary docs can be perceived as boilerplate duplication (some people wouldn't like having that)
* unit test summary docs are not that common from my experience
* the workaround is to copy it manually afterwards

Let's keep it as it is for now and eventually improve if the community wish so.

**Comment by aolszowka** [comment]
Sounds good to me; as everything is checked in can I just close out this issue?

**Comment by wachulski** [comment]
@aolszowka If you only could edit this PR and update source branch, ie. aolszowka/comment-and-region-truncation . As now it tells it wants to merge `14 commits into wachulski:master from unknown repository` and it should be 7 commits into wachulski:master from aolszowka/comment-and-region-truncation

We'd benefit from retaining the discussion history for future reference.

The minute you're done, I'll merge and release 1.1.1

**Comment by wachulski** [comment]
Ok, let me recreate the PR and link to this conversation. No worries.

When it comes to branching strategy, yes, I admit it takes time to make SVN->GIT switch.

> When contributing always first CLONE the master then NEVER commit directly to that;

Well, you actually can, providing it's not long-lasting (little changes, to be merged soon) and you make use of (intermediate level) rebasing technique.

In general merges are usually safe providing you sync with the base branch (master) often.

In short, these 2 tips: sync often, learn `rebase`. But anyway, a piece of solid work by you @aolszowka . Thanks.
